website/addons/box/model.py

```diff
@@ -0,0 +1,340 @@
+
+    #: Information saved at the time of registration
+    #: Note: This is unused right now
+    registration_data = fields.DictionaryField()
+
+
+    @property
+    def display_name(self):
+        return '{0}: {1}'.format(self.config.full_name, self.folder)
+
+    @property
+    def has_auth(self):
+        """Whether an access token is associated with this node."""
+        return bool(self.user_settings and self.user_settings.has_auth and refresh_creds_if_necessary(self.user_settings))
+
+    def set_folder(self, folder, folder_id, auth):
+        self.folder = folder
+        self.folder_id = folder_id
+        # Add log to node
+        nodelogger = BoxNodeLogger(node=self.owner, auth=auth)
+        nodelogger.log(action="folder_selected", save=True)
+
+    def set_user_auth(self, user_settings):
+        """Import a user's Box authentication and create a NodeLog.
+
+        :param BoxUserSettings user_settings: The user settings to link.
+        """
+        self.user_settings = user_settings
+        nodelogger = BoxNodeLogger(node=self.owner, auth=Auth(user_settings.owner))
+        nodelogger.log(action="node_authorized", save=True)
+
+    def find_or_create_file_guid(self, path):
+        return BoxFile.get_or_create(self.owner, clean_path(os.path.join(self.folder, path.lstrip('/'))))
+
+    # TODO: Is this used? If not, remove this and perhaps remove the 'deleted' field
+    def delete(self, save=True):
+        self.deauthorize(add_log=False)
+        super(BoxNodeSettings, self).delete(save)
+
+    def deauthorize(self, auth=None, add_log=True):
+        """Remove user authorization from this node and log the event."""
+        node = self.owner
+        folder = self.folder
+
+        self.folder = None
+        self.folder_id = None
+        self.user_settings = None
+
+        if add_log:
+            extra = {'folder': folder}
+            nodelogger = BoxNodeLogger(node=node, auth=auth)
+            nodelogger.log(action="node_deauthorized", extra=extra, save=True)
+
+    def serialize_waterbutler_credentials(self):
+        if not self.has_auth:
+            raise exceptions.AddonError('Addon is not authorized')
+        return {'token': self.user_settings.access_token}
+
+    def serialize_waterbutler_settings(self):
+        if not self.folder_id:
+            raise exceptions.AddonError('Folder is not configured')
+        return {'folder': self.folder_id}
+
+    def create_waterbutler_log(self, auth, action, metadata):
+        cleaned_path = clean_path(os.path.join(self.folder, metadata['path']))
+        self.owner.add_log(
+            'box_{0}'.format(action),
+            auth=auth,
+            params={
+                'project': self.owner.parent_id,
+                'node': self.owner._id,
+                'path': cleaned_path,
+                'folder': self.folder,
+                'urls': {
+                    'view': self.owner.web_url_for('box_view_file', path=cleaned_path),
```

Comment:
In as of 0.25 you can use `self.owner.web_url_for('addon_download_or_view_file', provider='box', action='view/download', path=...)`